File: Structural/Proxy/src/examplejava/medialibrary/YouTubeClass.java
package examplejava.medialibrary;

import java.util.HashMap;

public class YouTubeClass implements YouTubeLibrary {
    @Override
    public HashMap<String, Video> popularVideos() {
        connectToServer("http://www.youtube.com");
        return getRandomVideos();
    }

    @Override
    public Video getVideo(String videoId) {
        connectToServer("http://www.youtube.com/" + videoId);
        return getSomeVideo(videoId);
    }

    // -----------------------------------------------------------------------
    // Fake methods to simulate network activity. They as slow as a real life.
    private int random(int min, int max) {
        return min + (int) (Math.random() * ((max - min) + 1));
    }

    private void experienceNetWorkLatency() {
        int latency = random(5, 10);
        for (int i = 0; i < latency; i++) {
            try {
                Thread.sleep(100);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    private void connectToServer(String server) {
        System.out.println("Connecting to " + server + "...");
        experienceNetWorkLatency();
        System.out.println("Connected!");
    }

    private HashMap<String, Video> getRandomVideos() {
        System.out.println("Downloading populars");
        experienceNetWorkLatency();
        HashMap<String, Video> hmap = new HashMap<String, Video>();
        hmap.put("catzzzzzzzzz", new Video("sadgahasgdas", "Catzzzz.avi"));
        hmap.put("mkafksangasj", new Video("mkafksangasj", "Dog play with ball.mp4"));
        hmap.put("dancesvideoo", new Video("asdfas3ffasd", "Dancing video.mpq"));
        hmap.put("dlsdk5jfslaf", new Video("dlsdk5jfslaf", "Barcelona vs RealM.mov"));
        hmap.put("3sdfgsd1j333", new Video("3sdfgsd1j333", "Programing lesson#1.avi"));

        System.out.print("Done!" + "\n");
        return hmap;
    }

    private Video getSomeVideo(String videoId) {
        System.out.print("Downloading video... ");
        experienceNetWorkLatency();
        Video video = new Video(videoId, "Some video title");
        System.out.print("Done!" + "\n");
        return video;
    }
}

File: Structural/Proxy/src/examplekotlin/medialibrary/YouTubeClass.kt
package examplekotlin.medialibrary

import java.util.HashMap

class YouTubeClass : YouTubeLibrary {
    override fun popularVideos(): HashMap<String, Video> {
        connectToServer("http://www.youtube.com")
        return getRandomVideos()
    }

    override fun getVideo(videoId: String): Video {
        connectToServer("http://www.youtube.com/$videoId")
        return getSomeVideo(videoId)
    }

    // -----------------------------------------------------------------------
    // Fake methods to simulate network activity. They as slow as a real life.

    private fun random(min: Int, max: Int) = min + (Math.random() * (max - min + 1)).toInt()

    private fun experienceNetWorkLatency() {
        val latency = random(5, 10)
        for (i in 0..latency) {
            try {
                Thread.sleep(100)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
    }

    private fun connectToServer(server: String) {
        println("Connecting to $server...")
        experienceNetWorkLatency()
        println("Connected!")
    }

    private fun getRandomVideos(): java.util.HashMap<String, Video> {
        println("Downloading populars")
        experienceNetWorkLatency()
        val hmap = java.util.HashMap<String, Video>()
        hmap["catzzzzzzzzz"] = Video("sadgahasgdas", "Catzzzz.avi")
        hmap["mkafksangasj"] = Video("mkafksangasj", "Dog play with ball.mp4")
        hmap["dancesvideoo"] = Video("asdfas3ffasd", "Dancing video.mpq")
        hmap["dlsdk5jfslaf"] = Video("dlsdk5jfslaf", "Barcelona vs RealM.mov")
        hmap["3sdfgsd1j333"] = Video("3sdfgsd1j333", "Programing lesson#1.avi")
        print("""
    Done!
    
    """.trimIndent())
        return hmap
    }

    private fun getSomeVideo(videoId: String): Video {
        println("Downloading video")
        experienceNetWorkLatency()
        val video = Video(videoId, "Some video title")
        println("Done!" + "\n")
        return video
    }
}